easybuild/easyconfigs/c/CUDAcompat/CUDAcompat-11.6.eb

```diff
@@ -0,0 +1,33 @@
+name = 'CUDAcompat'
+version = '11.6'
+nv_version = '510.73.08'
```

Comment:
Should this be added as a `versionsuffix`? There are 5 `11.4` releases of the CUDAcompat libraries (for example).

Reply:
When going through the official process at https://www.nvidia.com/Download/index.aspx?lang=en-us I only get 1.

So I wouldn't litter the module name yet and do so when and if the need arises. See Python: No suffix as long is we only use 1/the default version

Reply:
Take a look at the compat libraries from https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/ and you can see all the driver releases for each minor version.

Given you know you have a uniqueness issue, I still think it is better to include.

Reply:
The driver versions do seem to map to a patch version though, so you could include that instead...but from what I can see the map from patch version to driver version can only be known by peeping at something like https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/ and figuring it out

Reply:
It seems that the runfile downloads are only there for the latest version, not the earlier ones. Hence on the runfile download page you only get 1 version.
It may be worth doing the Java approach and have 1 alias package for 11.6 and individual packages for the versions with the 11.6 one pointing to the latest.
IMO we can do that later when we decide to add another EC for the same CUDA version, but that would be an enhancement I don't think it is worth the work right now.

Reply:
But doesn't that mean that if a new version is released then the existing easyconfig ceases to function since it can't download sources?

Reply:
I'll go inspect the rpm and deb files to see if the libraries inside have the same checksum, if they do you can just take the deb package, unpack it with `ar` from binutils and `tar` and you'll still have something reproducible.

Reply:
Ok, to me it looks like you could directly use any arch appropriate package (and a `.deb` package is easiest for us to unpack):
```
alanc@~/compat$ find ./ -name libcuda.so.515.48.07 | xargs -i md5sum {}

febc658d85031c4f86bfdeb9e5e474d6  ./debian/usr/local/cuda-11.7/compat/libcuda.so.515.48.07

febc658d85031c4f86bfdeb9e5e474d6  ./rhel8/usr/local/cuda-11.7/compat/libcuda.so.515.48.07

febc658d85031c4f86bfdeb9e5e474d6  ./rhel9/usr/local/cuda-11.7/compat/libcuda.so.515.48.07

febc658d85031c4f86bfdeb9e5e474d6  ./opensuse/usr/local/cuda-11.7/compat/libcuda.so.515.48.07

febc658d85031c4f86bfdeb9e5e474d6  ./ubuntu/usr/local/cuda-11.7/compat/libcuda.so.515.48.07

alanc@~/compat$ find ./ -name libnvidia-nvvm.so.515.48.07 | xargs -i md5sum {}

f7bb01d27e8e78bed0b2adfdde70106a  ./debian/usr/local/cuda-11.7/compat/libnvidia-nvvm.so.515.48.07

f7bb01d27e8e78bed0b2adfdde70106a  ./rhel8/usr/local/cuda-11.7/compat/libnvidia-nvvm.so.515.48.07

f7bb01d27e8e78bed0b2adfdde70106a  ./rhel9/usr/local/cuda-11.7/compat/libnvidia-nvvm.so.515.48.07

f7bb01d27e8e78bed0b2adfdde70106a  ./opensuse/usr/local/cuda-11.7/compat/libnvidia-nvvm.so.515.48.07

f7bb01d27e8e78bed0b2adfdde70106a  ./ubuntu/usr/local/cuda-11.7/compat/libnvidia-nvvm.so.515.48.07

alanc@~/compat$ find ./ -name libnvidia-ptxjitcompiler.so.515.48.07 | xargs -i md5sum {}

23200efea3af20f49351c1fc92ee8c99  ./debian/usr/local/cuda-11.7/compat/libnvidia-ptxjitcompiler.so.515.48.07

23200efea3af20f49351c1fc92ee8c99  ./rhel8/usr/local/cuda-11.7/compat/libnvidia-ptxjitcompiler.so.515.48.07

23200efea3af20f49351c1fc92ee8c99  ./rhel9/usr/local/cuda-11.7/compat/libnvidia-ptxjitcompiler.so.515.48.07

23200efea3af20f49351c1fc92ee8c99  ./opensuse/usr/local/cuda-11.7/compat/libnvidia-ptxjitcompiler.so.515.48.07

23200efea3af20f49351c1fc92ee8c99  ./ubuntu/usr/local/cuda-11.7/compat/libnvidia-ptxjitcompiler.so.515.48.07
```

Reply:
Ah, Power is only available as an rpm, I guess we will have to use that.

Reply:
`p7zip` (available in GCCcore) can be used to extract an rpm, but it does not retain symlinks:
```
7z x cuda-compat-11-7-515.48.07-1.ppc64le.rpm 
7z x cuda-compat-11-7-515.48.07-1.ppc64le.cpio  
```

Reply:
> But doesn't that mean that if a new version is released then the existing easyconfig ceases to function since it can't download sources?

Seems the other sources are still there just not listed: E.g. https://us.download.nvidia.com/tesla/510.47.03/NVIDIA-Linux-x86_64-510.47.03.run

So I'd stick with the runfiles as that is what is documented and hence can be expected to work (and continue to work)

As for the nv_version: I still think that IF we want to add another version with the same CUDA version then we add a wrapper EC, rename the existing ECs and add the new one (the latter 2 with suffixes)

As that is a non-breaking change we can do later easily, I'd really do that later and first keep those here as-is for testing purposes. Maybe we later find some issue or so which may require a different approach. Especially as these ECs require a new EasyBlock I'd keep them as simple as possible (first)

Reply:
> > But doesn't that mean that if a new version is released then the existing easyconfig ceases to function since it can't download sources?
> 
> Seems the other sources are still there just not listed: E.g. https://us.download.nvidia.com/tesla/510.47.03/NVIDIA-Linux-x86_64-510.47.03.run

That doesn't seem to be true in all cases: https://us.download.nvidia.com/tesla/515.43.04/NVIDIA-Linux-x86_64-515.43.04.run (the previous driver release) . I guess that is because it is a feature branch? If these kinds of restrictions exist then we have a problem.

> So I'd stick with the runfiles as that is what is documented and hence can be expected to work (and continue to work)

It's a 300MB download instead of a 30MB download. If we can't guarantee that the sources are going to be available (which we seem to be able to do with the rpm files) I'm not so convinced by the fact that the approach is documented.

> 
> As for the nv_version: I still think that IF we want to add another version with the same CUDA version then we add a wrapper EC, rename the existing ECs and add the new one (the latter 2 with suffixes)
>
> As that is a non-breaking change we can do later easily, I'd really do that later and first keep those here as-is for testing purposes. Maybe we later find some issue or so which may require a different approach. Especially as these ECs require a new EasyBlock I'd keep them as simple as possible (first)

I disagree here, if you want to push forward with testing these then the name should retain uniqueness now. The Java-like behaviour is the thing that can be added later (and indeed, at that point you might as well just use `CUDAcompat/11` and drop the minor version).

Reply:
Hmm, that driver version is available at https://es.download.nvidia.com/XFree86/Linux-x86_64/515.43.04/NVIDIA-Linux-x86_64-515.43.04.run , there must be some limitiations on what falls under "tesla"

Reply:
> It's a 300MB download instead of a 30MB download. If we can't guarantee that the sources are going to be available (which we seem to be able to do with the rpm files) I'm not so convinced by the fact that the approach is documented.

I don't see 270MB as an actual issue... That's small nowadays. But yes, the seemingly longer availability of the repo-stuff is a valid point. We could even support both in the easyblock by checking the extension.
My issue with those is that https://developer.download.nvidia.com/compute/cuda/repos/ lists a lot of downloads. For me it isn't clear which one we can/should take or if any is fine.

> I disagree here, if you want to push forward with testing these then the name should retain uniqueness now. The Java-like behaviour is the thing that can be added later (and indeed, at that point you might as well just use CUDAcompat/11 and drop the minor version).

No we'll always need a 11.6, 11.7 etc. due to the different driver requirements as listed on https://docs.nvidia.com/deploy/cuda-compatibility/index.html#manually-installing-from-runfile So the current name IS unique. You want to have a "CUDAcompat/11.6" and appending the driver version will only be confusing (currently)
So we either need to do the Java thing now adding complexity or when required. In both cases a `eb CudaCompat-11.6.eb --rebuild -r` will update to whatever we settle on later.

Reply:
> I don't see 270MB as an actual issue... That's small nowadays. But yes, the seemingly longer availability of the repo-stuff is a valid point. We could even support both in the easyblock by checking the extension. My issue with those is that https://developer.download.nvidia.com/compute/cuda/repos/ lists a lot of downloads. For me it isn't clear which one we can/should take or if any is fine.

I did some testing in https://github.com/easybuilders/easybuild-easyconfigs/pull/15892#discussion_r929766927 and it doesn't seem to matter. The major thing is only RHEL seem to have all archs so the RPMs in RHEL8 seem like the best place to get them. 

> 
> No we'll always need a 11.6, 11.7 etc. due to the different driver requirements as listed on https://docs.nvidia.com/deploy/cuda-compatibility/index.html#manually-installing-from-runfile So the current name IS unique. You want to have a "CUDAcompat/11.6" and appending the driver version will only be confusing (currently) So we either need to do the Java thing now adding complexity or when required. In both cases a `eb CudaCompat-11.6.eb --rebuild -r` will update to whatever we settle on later.

That doesn't entirely capture the issue, and I don't think that's quite true. You will always want the latest possible version of the compat libraries (and the latest version is backwards compatible with all previous 11.x versions) so I don't see a good reason to use an `11.6` compat library if you have an `11.7` available. If you try to use an older compat library with a newer driver things will fail (and this could easily happen if you stick with an `11.x` approach). You could mitifate for this by only conditionally adding the compat libraries to `LD_LIBRARY_PATH` based on the current installed driver.

Reply:
> So the current name IS unique. You want to have a "CUDAcompat/11.6" and appending the driver version will only be confusing (currently)

Sorry but this is clearly not true. For `CUDAcompat/11.4` you could have 5 different  `nv_version`s (see https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/) which give the same final module file, without the driver suffix you have a uniqueness problem.

Reply:
I get your point on the 418 driver and the CUDA/11.7 compat library (which is not compatible)...but I don't foresee this as a problem, if CUDAcompat/11 was updated to require an 11.7 flavour then you'd have to install that dependency, the easyblock should be smart enough to catch this incompatibility (through sanity checks).

Reply:
> * The different driver versions of the compat package do not matter for compatibility (given a fixed major&minor version)
> 
> Anything wrong?

Yes I agree with that, the only point being the driver version of the compat package. I suspect the version _does_ matter if you are _not_ on the driver branches listed in the tables at https://docs.nvidia.com/deploy/cuda-compatibility/index.html#use-the-right-compat-package, in which case whether the compat libraries will work for you is probably arbitrary.

> Hence I think a `CUDAcompat/11` package isn't very useful to have in EasyBuild: You won't know if it is compatible with your driver. So we need to include the minor versions in the ECs so site admins can install the EC which is compatible with the (currently) installed driver.

I think that depends on clever the underlying easyblock is. To install/update `CUDAcompat/11` (a modulerc) you'd need to (ultimately) install  `CUDAcompat/11.X-DRIVER_VERSION`, if the easyblock actually checks compatibility for you then this will fail gracefully with a helpful error message. I get your point though, but I think you answer it yourself below.

> Of course we could stack this:
> 
> * CUDAcompat/11 depends on the latest CUDAcompat/11.x
> * CUDAcompat/11.x depends on the latest CUDAcompat/11.x-y
> 
> But we need the "middle layer".

I think that this could work. Most people would prefer to have `CUDAcompat/11` that is dynamic and updates to the latest available compat package. 

> > For CUDAcompat/11.4 you could have 5 different nv_versions
> 
> Yes, "could". But currently there is only 1 EC. And as (according to the above) we will always need a `CUDAcompat/11.4` it is easy to change the existing EC to have a version suffix and add an alias EC.

We always start with one ec but it is rare we stop there. If we already know we may need to change things, I think it is better we start off with the right path rather than change the behaviour after the fact.

> 
> So my point is that **currently** the version suffix is not required and not adding it doesn't prevent us from adding it later **when** it is required. Also not adding it now, means having only 2 new ECs instead of 4. I really want to be able to add `CUDAcompat/11.6` as a dependency to CUDA in site specific hooks.

I think using a modulerc for `CUDAcompat/11.6` is a more robust and reproducible solution. I'll test it out today.

Reply:
Ok, I tested using a modulerc to wrap access to the specific `CUDAcompat` version and this seems to work without issue. This was with a trivial change to the ec in this PR:
```
name = 'CUDAcompat'
version = '11.6'
versionsuffix = "-D%s" % nv_version
```
and then the creation of two new (trivial) easyconfigs, `CUDAcompat-11.6.eb`:
```
easyblock = 'ModuleRC'

name = 'CUDAcompat'
version = '11.6'

homepage = 'https://docs.nvidia.com/deploy/cuda-compatibility/index.html'
description = """Using the CUDA Forward Compatibility package,
 system administrators can run applications built using a newer toolkit
 even when an older driver that does not satisfy the minimum required driver version
 is installed on the system.
 This forward compatibility allows the CUDA deployments in data centers and enterprises
 to benefit from the faster release cadence and the latest features and performance of CUDA Toolkit.
"""

toolchain = SYSTEM

dependencies = [(name, "11.6", "-D510.73.08")]

moduleclass = 'system'
```
and `CUDAcompat-11.eb`
```
easyblock = 'ModuleRC'

name = 'CUDAcompat'
version = '11'

homepage = 'https://docs.nvidia.com/deploy/cuda-compatibility/index.html'
description = """Using the CUDA Forward Compatibility package,
 system administrators can run applications built using a newer toolkit
 even when an older driver that does not satisfy the minimum required driver version
 is installed on the system.
 This forward compatibility allows the CUDA deployments in data centers and enterprises
 to benefit from the faster release cadence and the latest features and performance of CUDA Toolkit.
"""

toolchain = SYSTEM

dependencies = [(name, "11.6")]

moduleclass = 'system'
```

Reply:
> Most people would prefer to have CUDAcompat/11 that is dynamic and updates to the latest available compat package.

Why do you think that? If at our cluster someone was to install that EC it would fail right away as the driver is not compatible with 11.7. So choosing 11.6 directly is the better choice (and the only option)

Anyway added the 11.6, 11.7 and 11 ECs as aliases so everyone can choose. Also added the version check to the easyblock. We could even omit the CUDA version from the "lowest level" ECs as they are already versioned: By the driver version. But keeping it allows the sanity check to work and users/admins to easily see which one applies. Made them hidden though as you should rather load CUDAcompat/11.6 instead of an individual driver version of the CUDA version.

Reply:
> > Most people would prefer to have CUDAcompat/11 that is dynamic and updates to the latest available compat package.
> 
> Why do you think that? If at our cluster someone was to install that EC it would fail right away as the driver is not compatible with 11.7. So choosing 11.6 directly is the better choice (and the only option)

That's kind of my point, you're making the assumption that the driver installed will never be updated. That might be true in your case, but it's unlikely to be true for everyone. If the driver gets updated to an 11.7+ driver then all your modules that depend on `11.6` will start failing.

What the modulerc actually points to is a site choice. The two modulerc easyconfigs would likely have to be edited in a situation like yours anyway based on when when your driver is EOL. If the CUDAcompat easyblock was robust enough to actually check if the compat libs worked then this would be a non-issue since the primary installation would fail anyway (but as discussed in the easyblock PR that is easier said than done).

In your particular case, you would choose `11` to point to `11.6`, if you _had_ a choice obviously you would want the latest and greatest.

Reply:
> That's kind of my point, you're making the assumption that the driver installed will never be updated. That might be true in your case, but it's unlikely to be true for everyone. If the driver gets updated to an 11.7+ driver then all your modules that depend on 11.6 will start failing.

Understood. So adding both layers as discussed earlier makes sense. Site admins now have the full choice if they want the latest 11 package, the latest 11.x package or a specific version.

Anything else missing?